File: Projects/FlutterPerfLab/lib/modules/videotest/video_test_view_model.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:share_plus/share_plus.dart';

import 'package:flutterperflab/core/managers/performance_manager.dart';
import 'package:flutterperflab/core/managers/export_manager.dart';
import 'package:flutterperflab/core/engine/video_engine.dart';

/// A centralized controller managing the lifecycle of high-frequency video decoding benchmarks.
class VideoTestViewModel extends ChangeNotifier {
  // MARK: - Reactive UI State

  bool isTestingActive = false;
  bool isFullScreen = false;
  bool isReportReady = false;

  // CORE DEPENDENCIES
  final PerformanceManager _perfManager = PerformanceManager.shared;
  final VideoEngine engine = VideoEngine.shared;

  VideoTestViewModel() {
    setupBindings();
  }

  void setupBindings() {
    engine.addListener(_onEngineStateChanged);
  }

  void _onEngineStateChanged() {
    // Safely terminates the benchmark ONLY when the underlying native decoder
    // explicitly broadcasts an End-Of-File state.
    if (engine.hasEnded && isTestingActive) {
      if (kDebugMode) {
        print(
          "[Benchmark] Video EOF Detected -> Finalizing benchmark and generating report.",
        );
      }

      // Triggers stopTest() to finalize the report and display the Export button.
      stopTest(isFinished: true);
    }
  }

  // MARK: - Hardware Resource Allocation

  Future<void> prepareVideo() async {
    isReportReady = false;
    notifyListeners();

    await engine.prepareVideo(name: "test_video_1080p", ext: "mp4");
  }

  // MARK: - Benchmark Initialization

  Future<void> startTest() async {
    if (isTestingActive) return;

    if (engine.controller == null) {
      await prepareVideo();
    }

    isReportReady = false;
    isFullScreen = true;
    isTestingActive = true;
    notifyListeners();

    _perfManager.startMonitoring();
    await engine.playFromStart();
  }

  // MARK: - Data Persistence & Finalization

  Future<void> stopTest({required bool isFinished}) async {
    if (!isTestingActive && !isFinished) return;

    isTestingActive = false;
    isFullScreen = false;
    notifyListeners();

    await engine.stop();
    _perfManager.stopMonitoring();

    // Restores default UI state and orientation upon termination.
    await SystemChrome.setPreferredOrientations([
      DeviceOrientation.portraitUp,
      DeviceOrientation.portraitDown,
    ]);
    await SystemChrome.setEnabledSystemUIMode(SystemUiMode.edgeToEdge);

    final String lastTestSuffix = isFinished ? "Complete" : "Partial";

    // Asynchronous I/O Serialization
    try {
      final file = await ExportManager.shared.generateCSV(
        logs: _perfManager.currentLogs,
        testName: "Video_FullScreen_Flutter_$lastTestSuffix",
      );

      if (file != null) {
        isReportReady = true;
        notifyListeners(); // Exposes the Export button on the UI.
      }
    } catch (e) {
      if (kDebugMode) {
        print("Export Failed: $e");
      }
    }
  }

  // MARK: - Inter-Process Communication

  Future<void> shareResults() async {
    if (!isReportReady) return;

    try {
      final file = await ExportManager.shared.generateCSV(
        logs: _perfManager.currentLogs,
        testName: "Video_FullScreen_Flutter_ManualExport",
      );

      if (file != null) {
        final xFile = XFile(file.path);
        await Share.shareXFiles([xFile]);
      }
    } catch (_) {}
  }

  // MARK: - Hardware Orientation Override

  Future<void> forceRotation({required bool isLandscape}) async {
    if (isLandscape) {
      await SystemChrome.setPreferredOrientations([
        DeviceOrientation.landscapeRight,
        DeviceOrientation.landscapeLeft,
      ]);
      await SystemChrome.setEnabledSystemUIMode(SystemUiMode.immersiveSticky);
    } else {
      await SystemChrome.setPreferredOrientations([
        DeviceOrientation.portraitUp,
        DeviceOrientation.portraitDown,
      ]);
      await SystemChrome.setEnabledSystemUIMode(SystemUiMode.edgeToEdge);
    }
  }

  @override
  void dispose() {
    engine.removeListener(_onEngineStateChanged);
    engine.release();
    super.dispose();
  }
}

File: Projects/KotlinMultiPlatformPerfLab/composeApp/src/commonMain/kotlin/com/tez/perflab/ui/features/videotest/VideoTestViewModel.kt
package com.tez.perflab.ui.features.videotest

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tez.perflab.engines.VideoEngine
import com.tez.perflab.managers.ExportManager
import com.tez.perflab.managers.PerformanceManager
import com.tez.perflab.managers.PlatformContext
import com.tez.perflab.managers.lockScreenOrientation
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * A cross-platform, lifecycle-aware controller managing full-screen video playback benchmarks.
 * Centralizes orientation lockdowns, hardware telemetry initiation, and deterministic
 * serialization into a single unified layer.
 */
class VideoTestViewModel : ViewModel() {

    val engine = VideoEngine
    private val perfManager = PerformanceManager

    // Caches context strictly to execute background CSV persistence, restore OS-level
    // screen orientation locks, and invoke native Share Intent APIs.
    private var currentContext: PlatformContext? = null

    // Utilizing primitive types for internal execution tracking to minimize state allocation overhead.
    private var isTestingActive = false

    // Retains the execution state for accurate telemetry serialization nomenclature.
    private var lastTestSuffix = "Partial"

    // MARK: - Reactive UI State

    private val _isFullScreen = MutableStateFlow(false)
    val isFullScreen: StateFlow<Boolean> = _isFullScreen.asStateFlow()

    private val _isReportReady = MutableStateFlow(false)
    val isReportReady: StateFlow<Boolean> = _isReportReady.asStateFlow()

    init {
        setupBindings()
    }

    private fun setupBindings() {
        viewModelScope.launch {
            engine.hasEnded.collect { ended ->
                if (ended && isTestingActive) {
                    println("[Telemetry] Video EOF Detected -> Gracefully terminating the active benchmark suite.")
                    stopTest(isFinished = true)
                }
            }
        }
    }

    fun prepareVideo(context: PlatformContext) {
        this.currentContext = context
        _isReportReady.value = false
        engine.prepareVideo(context, "test_video_1080p", "mp4")
    }

    fun startTest(context: PlatformContext) {
        this.currentContext = context
        if (engine.getPlayer() == null) {
            prepareVideo(context)
        }

        _isReportReady.value = false

        // 1. Hardware Override: Lock to Landscape
        lockScreenOrientation(context, isLandscape = true)

        _isFullScreen.value = true
        isTestingActive = true

        // 2. Initiate Hardware Telemetry & Playback
        perfManager.startMonitoring(context)
        engine.playFromStart()
    }

    fun stopTest(isFinished: Boolean = false) {
        if (!isTestingActive) return
        val context = currentContext ?: return

        isTestingActive = false
        _isFullScreen.value = false

        // 1. Restore OS UX Constraints
        lockScreenOrientation(context, isLandscape = false)

        // 2. Terminate Hardware Decoders and Telemetry Trackers
        engine.stop()
        perfManager.stopMonitoring()

        // Release the context reference immediately post-execution
        // to prevent retention leaks prior to the export phase.
        currentContext = null

        // Caches the terminal state identifier for subsequent I/O operations.
        lastTestSuffix = if (isFinished) "Complete" else "Partial"

        _isReportReady.value = true
    }

    fun shareResults(context: PlatformContext) {
        // Delegating file serialization to the I/O thread pool to prevent
        // CPU-bound thread starvation during heavy disk writes.
        viewModelScope.launch(Dispatchers.IO) {
            try {
                ExportManager.generateAndShareCSV(
                    logs = perfManager.currentLogs.value,
                    testName = "Video_FullScreen_KMP_$lastTestSuffix",
                    context = context
                )
            } catch (e: Exception) {
                println("[Export Error] ${e.message}")
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        engine.release()
        currentContext?.let { lockScreenOrientation(it, isLandscape = false) }
        currentContext = null
    }
}